Sources/NIO/EventLoop.swift

```diff
@@ -1125,6 +1152,19 @@ public final class MultiThreadedEventLoopGroup: EventLoopGroup {
             }
 
             handler(error)
+            self.shutdownLock.withLock {
+                guard case .closing(let callbacks) = self.runState else {
+                    fatalError("runState was \(self.runState), expected .closing")
+                }
+
+                for (q, handler) in callbacks {
+                    q.async {
+                        handler(error)
+                    }
+                }
+
+                self.runState = .closed(error)
+            }
```

Comment:
It would probably be better to perform the state transition and then create a copy of the callback list and zero it out, then drop the lock, then make the outcalls. Making arbitrary outcalls while holding locks is a good way to get deadlocks, which is what this patch is trying to save us from.

Reply:
I can make that change, though I don't think there's any risk for deadlocks here. The only thing that happens is that the callbacks get dispatched onto their assigned dispatch queue and no other locks will be obtained, but there's no downside to reducing the amount of time spent in the lock.